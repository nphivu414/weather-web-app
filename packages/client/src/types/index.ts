export interface Place {
  title: string;
  location_type: string;
  latt_long: string;
  woeid: number;
  distance?: number;
  consolidated_weather?: Weather[];
}

export interface Weather {
  id: string;
  weather_state_name: string;
  weather_state_abbr: string;
  wind_direction_compass: string;
  created: string;
  applicable_date: string;
  min_temp: number;
  max_temp: number;
  the_temp: number;
  wind_speed: number;
  wind_direction: number;
  air_pressure: number;
  humidity: number;
  visibility: number;
  predictability: number;
}

export interface ServiceError {
  statusCode: number;
  message: string;
}
