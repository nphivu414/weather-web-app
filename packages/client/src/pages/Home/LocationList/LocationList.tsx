import { BiChevronRight } from 'react-icons/bi';
import * as React from 'react';
import { Place } from 'types';

export type LocationListProps = {
  locations?: Place[] | null;
  isFetched?: boolean;
  onItemClick?: (location: Place) => void;
};

export const handleOnItemClick = (location: Place, onItemClick?: (location: Place) => void) => {
  return () => {
    onItemClick && onItemClick(location);
  };
};

const LocationList: React.FC<LocationListProps> = ({ locations, isFetched, onItemClick }) => {
  if (!locations || locations.length === 0) {
    if (isFetched) {
      return <p className="text-center text-muted empty">No data</p>;
    }
    return <p className="text-center text-muted hint">Start typing to search for cities</p>;
  }

  return (
    <div className="list-group">
      {locations.map((item) => {
        const { woeid, title } = item;
        return (
          <button
            key={woeid}
            className="list-group-item list-group-item-action d-flex justify-content-between align-items-center"
            onClick={handleOnItemClick(item, onItemClick)}
          >
            {`${title}`}
            <BiChevronRight size="20px" />
          </button>
        );
      })}
    </div>
  );
};

export default LocationList;
