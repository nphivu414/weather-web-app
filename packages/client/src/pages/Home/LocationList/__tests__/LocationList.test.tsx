import * as React from 'react';
import { shallow } from 'enzyme';

import LocationList, { handleOnItemClick, LocationListProps } from '../LocationList';
import { Place } from 'types';

const props: LocationListProps = {
  locations: [
    {
      latt_long: '123-345',
      location_type: 'Test',
      title: 'Test',
      woeid: 1,
    },
  ],
  onItemClick: jest.fn(),
};

describe('LocationList', () => {
  it('should match a snapshot', () => {
    expect(shallow(<LocationList {...props} />)).toMatchSnapshot();
  });

  it('should show no data text', () => {
    const wrapper = shallow(<LocationList {...props} locations={null} isFetched={true} />);
    expect(wrapper.find('p.empty')).toHaveLength(1);
  });

  it('should show hint text', () => {
    const wrapper = shallow(<LocationList {...props} locations={null} isFetched={false} />);
    expect(wrapper.find('p.hint')).toHaveLength(1);
  });

  describe('handleOnItemClick', () => {
    it('should run correctly', () => {
      const mockLocation: Place = {
        latt_long: '123-345',
        location_type: 'Test',
        title: 'Test',
        woeid: 1,
      };
      const fn = handleOnItemClick(mockLocation, props.onItemClick);
      fn();
      expect(props.onItemClick).toBeCalledWith(mockLocation);
    });
  });
});
