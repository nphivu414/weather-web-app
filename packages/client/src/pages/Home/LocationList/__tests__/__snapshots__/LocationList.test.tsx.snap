// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationList should match a snapshot 1`] = `ShallowWrapper {}`;
